[JSC] Add parsing for deferred import proposal

takikawa opened the pull request:
#### 4e8dd496fc1158acdbcd5f20dbe67e9b13475386
<pre>
[JSC] Add parsing for deferred import proposal

<a href="https://bugs.webkit.org/show_bug.cgi?id=283133">https://bugs.webkit.org/show_bug.cgi?id=283133</a>

Reviewed by Yusuke Suzuki.

This patch adds initial parsing support for deferred
imports:
  * Adds feature flag for deferred imports
  * Adds syntax for `import defer` (no `import.defer` yet)
  * Adds parsing tests
  * Adds part of data representation for deferred
    imports relevant to the parser
  * `import defer` acts identical to plain `import` for now

The spec for the proposal is at:
  <a href="https://github.com/tc39/proposal-defer-import-eval/">https://github.com/tc39/proposal-defer-import-eval/</a>

* JSTests/stress/modules-syntax-error.js:
* JSTests/stress/modules-syntax-import-defer-error.js: Added.
(shouldThrow):
* JSTests/stress/modules-syntax-import-defer.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createImportDeclaration):
* Source/JavaScriptCore/parser/NodeConstructors.h:
(JSC::ImportDeclarationNode::ImportDeclarationNode):
* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser&lt;LexerType&gt;::parseImportDeclaration):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: <a href="https://commits.webkit.org/287419@main">https://commits.webkit.org/287419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9b34c918e4b9b08fa6b62cec2b9e12d1c5639bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58402 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83985 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30511 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67468 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6665 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62088 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19968 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52145 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72320 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42397 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49505 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26507 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28915 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72461 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70730 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26953 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85389 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78564 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6651 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4633 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70341 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6816 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68197 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69698 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17387 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13630 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12497 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100906 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6603 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24661 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6662 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9991 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (commented):
LGTM

Constellation commented on Source/JavaScriptCore/parser/ASTBuilder.h:
```diff
@@ -832,9 +832,9 @@ class ASTBuilder {
     }
 
-    StatementNode* createImportDeclaration(const JSTokenLocation& location, ImportSpecifierListNode* importSpecifierList, ModuleNameNode* moduleName, ImportAttributesListNode* importAttributesList)
+    StatementNode* createImportDeclaration(const JSTokenLocation& location, const ImportDeclarationNode::ImportType type, ImportSpecifierListNode* importSpecifierList, ModuleNameNode* moduleName, ImportAttributesListNode* importAttributesList)
```
We do not attach `const` for value parameter like this if it is not a reference.

takikawa replied:
Thanks, fixed.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3690,20 +3691,37 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseImportDeclara
+        if (Options::useImportDefer() && matchContextualKeyword(m_vm.propertyNames->deferKeyword)) {
+            // import defer NameSpaceImport FromClause ;
+            // import defer FromClause ;
+            JSTokenLocation deferLocation(tokenLocation());
```
Why not put this first before `(matchSpecIdentifier()` check?

takikawa replied:
Thanks for the suggestion, I refactored it to put it first. I also removed some code duplication (`parseImportClauseItem` does not need to be called here, it's enough to check we have `TIMES` and defer to existing code below).

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3690,11 +3691,23 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseImportDeclara
-    if (matchSpecIdentifier()) {
+    if (Options::useImportDefer() && matchContextualKeyword(m_vm.propertyNames->deferKeyword)) {
+        // import defer NameSpaceImport FromClause ;
+        // import defer FromClause ;
```
I cannot find this case from the proposal text. Can you specify the URL?

takikawa replied:
This case is parsing for a normal default import where the name is `defer`, rather than a syntax that's part of the deferred import proposal. I put in a case for it here because without peeking at the next token after `defer` (which I think we can't do right?) for `TIMES`, it's not possible to tell if we're parsing `import defer` or `import` with a named import `defer` until after consuming `defer`.

I realized though there is a bug here, in that we should also allow `import defer, * as name from "mod"` to parse.

Constellation replied:
Since `import defer FromClause` looks confusing, let's put `import defer NameSpaceImport FromClause ;` after `match(TIMES)` pass as a comment.

takikawa replied:
Done, I split up the comment and put them in the branches

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3690,11 +3691,23 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseImportDeclara
 
     bool isFinishedParsingImport = false;
-    if (matchSpecIdentifier()) {
+    if (Options::useImportDefer() && matchContextualKeyword(m_vm.propertyNames->deferKeyword)) {
```
I think this is not correct since if we have named-import with the name `defer`, we should not go to this path.

Constellation reviewed (changes requested):
Please add URL to the spec in the commit message.

Constellation commented on Source/JavaScriptCore/parser/Parser.cpp:
```diff
@@ -3690,11 +3691,26 @@ template <class TreeBuilder> TreeStatement Parser<LexerType>::parseImportDeclara
+            if (match(COMMA))
+                next();
+            else
+                isFinishedParsingImport = true;
```
This looks not correct, as it is not corresponding to what `parseImportClauseItem` is doing.
We should not expand the code here, and we should use `parseImportClauseItem` instead.
Please use savePoint and restoreSavePoint mechanism when `defer` is detected. And doing special thing only when you find `defer` and `*`, and otherwise, let's just restore and going to the normal path following.

takikawa replied:
Thanks for the feedback, just pushed a version that uses save points instead.

Constellation reviewed (approved):
r=me

takikawa committed:
[JSC] Add parsing for deferred import proposal

https://bugs.webkit.org/show_bug.cgi?id=283133

Reviewed by Yusuke Suzuki.

This patch adds initial parsing support for deferred
imports:
  * Adds feature flag for deferred imports
  * Adds syntax for `import defer` (no `import.defer` yet)
  * Adds parsing tests
  * Adds part of data representation for deferred
    imports relevant to the parser
  * `import defer` acts identical to plain `import` for now

The spec for the proposal is at:
  https://github.com/tc39/proposal-defer-import-eval/

* JSTests/stress/modules-syntax-error.js:
* JSTests/stress/modules-syntax-import-defer-error.js: Added.
(shouldThrow):
* JSTests/stress/modules-syntax-import-defer.js: Added.
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/parser/ASTBuilder.h:
(JSC::ASTBuilder::createImportDeclaration):
* Source/JavaScriptCore/parser/NodeConstructors.h:
(JSC::ImportDeclarationNode::ImportDeclarationNode):
* Source/JavaScriptCore/parser/Nodes.h:
* Source/JavaScriptCore/parser/Parser.cpp:
(JSC::Parser<LexerType>::parseImportDeclaration):
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/OptionsList.h:

Canonical link: https://commits.webkit.org/287419@main
